### DISCUSSION TITLE
DOM not updating in x-for loop after sorting using SortableJS

### DISCUSSION THREAD

**Discussion by danbramall** [alpinejs/alpine#4242] [1. Help]
I have an AlpineJS UI which allows an editor to add style rules or groups of style rules to a page (it's not really important what these are but I've kept the concept so as to differentiate from a simpler application).

After adding rules and rule groups you can sort them by dragging/dropping. It uses [SortableJS](http://sortablejs.github.io/Sortable/) for this.

The rules created are output in JSON format so they can be copied/pasted into a third-party application.

HTML:

```
<div class="configurator" x-data="ruleConfigurator()" x-init="init()">

  <template x-for="(rule, index) in rules" :key="index">
    <div class="rule-wrapper">
      <template x-if="rule.type === 'single'">
        <div class="rule rule--single">
          <div class="rule__cell rule__cell--title">
            <input type="text" placeholder="e.g. Heading 1" x-model="rule.title" @input="updateOutput">
          </div>
          <div class="rule__cell rule__cell--type">
            <select x-model="rule.category" @change="updateOutput">
              <option value="block">Block</option>
              <option value="inline">Inline</option>
              <option value="selector">Selector</option>
            </select>
          </div>
          <div class="rule__cell rule__cell--selector">
            <input type="text" placeholder="e.g. h1" x-model="rule.selector" @input="updateOutput">
          </div>
          <div class="rule__cell rule__cell--classes">
            <input type="text" placeholder="e.g. heading-primary" x-model="rule.classes" @input="updateOutput">
          </div>
          <button class="rule__delete" @click="removeRule(index)">X</button>
          <button class="rule__sort">Sort</button>
        </div>
      </template>
      <template x-if="rule.type === 'group'">
        <div class="rule rule--group">
          <div class="rule__cell rule__cell--group-title">
            <input type="text" placeholder="e.g. Headings" x-model="rule.title" @input="updateOutput">
          </div>

          <template x-for="(groupRule, groupRuleIndex) in rule.groupRules" :key="groupRuleIndex">
            <div class="rule__cell rule__cell--rules">
              <div class="rule rule--single">

                <div class="rule__cell rule__cell--title">
                  <input type="text" placeholder="title" x-model="groupRule.title" @input="updateOutput">
                </div>
                <div class="rule__cell rule__cell--type">
                  <select x-model="groupRule.category" @change="updateOutput">
                    <option value="block">Block</option>
                    <option value="inline">Inline</option>
                    <option value="selector">Selector</option>
                  </select>
                </div>
                <div class="rule__cell rule__cell--selector">
                  <input type="text" placeholder="selector" x-model="groupRule.selector" @input="updateOutput">
                </div>
                <div class="rule__cell rule__cell--classes">
                  <input type="text" placeholder="classes" x-model="groupRule.classes" @input="updateOutput">
                </div>
                <button class="rule__delete" @click="removeGroupRule(index, groupRuleIndex)">X</button>

              </div>
            </div>
          </template>
          <div class="add-rules">
            <button class="add-rules__btn btn js-add-style-to-group" @click="addRuleToGroup(index)">Add Style To Group</button>
          </div>

          <button class="rule__delete" @click="removeRule(index)">X</button>
          <button class="rule__sort">Sort</button>
  
        </div>
  
      </template>
    </div>
  </template>

  <div class="add-rules">
    <button @click="addRule('single')" class="add-rules__btn btn">Add Single Style</button>
    <button @click="addRule('group')" class="add-rules__btn btn">Add Style Group</button>  
  </div>

  <div class="output">
    <p>Output:</p>
    <textarea readonly x-text="output" class="output-ui__textarea js-output"></textarea>
  </div>
  
</div>

  </div>
```

Typescript:

```
import Alpine from 'https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/+esm';
import Sortable, { SortableEvent } from 'https://cdn.jsdelivr.net/npm/sortablejs@1.15.2/+esm';

window.Alpine = Alpine;

interface BaseRule {
  title: string;
  category: 'block' | 'inline' | 'selector';
  selector: string;
  classes: string;
}

interface SingleRule extends BaseRule {
  type: 'single';
}

interface GroupRule extends BaseRule {}

interface Group extends BaseRule {
  type: 'group';
  groupRules: GroupRule[];
}

type Rule = SingleRule | Group;

function ruleConfigurator() {
  return {
    rules: [] as Rule[],
    output: '',

    init() {
      this.initSortable();
      this.updateOutput();
    },

    initSortable() {
      const rulesContainer = document.querySelector('.configurator') as HTMLElement;
      Sortable.create(rulesContainer, {
        handle: '.rule__sort',
        onEnd: (event: SortableEvent) => {
          console.log(event);
          const movedRule = this.rules.splice(event.oldIndex!, 1)[0];
          this.rules.splice(event.newIndex!, 0, movedRule);
          this.updateOutput();
        }
      });
    },

    addRule(type: 'single' | 'group') {
      if (type === 'single') {
        this.rules.push({
          type: 'single',
          title: '',
          category: 'block',
          selector: '',
          classes: ''
        } as SingleRule);
      } else {
        this.rules.push({
          type: 'group',
          title: '',
          category: 'block',
          selector: '',
          classes: '',
          groupRules: [
            {
              title: '',
              category: 'block',
              selector: '',
              classes: ''
            }
          ]
        } as Group);
      }
      this.updateOutput();
    },

    addRuleToGroup(ruleIndex: number) {
      const rule = this.rules[ruleIndex];
      if (rule.type === 'group') {
        rule.groupRules.push({
          title: '',
          category: 'block',
          selector: '',
          classes: ''
        });
        this.updateOutput();
      }
    },

    removeRule(index: number) {
      this.rules.splice(index, 1);
      this.updateOutput();
    },

    removeGroupRule(ruleIndex: number, groupRuleIndex: number) {
      const rule = this.rules[ruleIndex];
      if (rule.type === 'group') {
        rule.groupRules.splice(groupRuleIndex, 1);
        this.updateOutput();
      }
    },

    updateOutput() {
      this.output = JSON.stringify(this.rules, null, 2);
    }
  };
}

document.addEventListener('alpine:init', () => {
  Alpine.data('ruleConfigurator', ruleConfigurator);
});

Alpine.start();
```

[There's a CodePen here.](https://codepen.io/danb1/pen/PovbBgO?editors=1010)

The issue I'm facing is with sorting. The code in the `onEnd` callback updates the rules object so that the JSON output shows the rules in the correct order, however the UI does not update — after dragging a rule to a new position the UI reverts back to how it was originally. If I comment out the stuff in the `onEnd` callback the UI sorting works correctly but the JSON object doesn't change hence is now in the wrong order.

I'd like the two things to be in sync so that the UI updates correctly when sorted and the JSON is also updated to reflect the new order.

Can anyone tell why this is happening and how it can be fixed?

Many thanks!

**Comment by ekwoka** [comment]
Is there a reason why you can't use the Alpine Sort Core Plugin? 

Using index as the key will also cause tons of problems if you're sorting them, since alpine would be spending a lot of time overwriting and rewriting and not keep things synced well. You should add unique ids.

Looked closer, there is also just an issue that you are implicitely updating the output, but if you use a getter, it will work fine.

[Here is a demo](https://awesomealpine.com/play/?coreplugins=0&ts=1&tw=1&html=PGRpdgogIGNsYXNzPSJjb25maWd1cmF0b3IiCiAgeC1kYXRhPSJydWxlQ29uZmlndXJhdG9yKCkiCiAgeC1pbml0PSJpbml0KCkiCj4KICA8dGVtcGxhdGUKICAgIHgtZm9yPSJydWxlIG9mIHJ1bGVzIgogICAgOmtleT0icnVsZS5pZCIKICA-CiAgICA8ZGl2IGNsYXNzPSJydWxlLXdyYXBwZXIiPgogICAgICA8dGVtcGxhdGUgeC1pZj0icnVsZS50eXBlID09PSAnc2luZ2xlJyI-CiAgICAgICAgPGRpdiBjbGFzcz0icnVsZSBydWxlLS1zaW5nbGUiPgogICAgICAgICAgPGRpdiBjbGFzcz0icnVsZV9fY2VsbCBydWxlX19jZWxsLS10aXRsZSI-CiAgICAgICAgICAgIDxpbnB1dAogICAgICAgICAgICAgIHR5cGU9InRleHQiCiAgICAgICAgICAgICAgcGxhY2Vob2xkZXI9ImUuZy4gSGVhZGluZyAxIgogICAgICAgICAgICAgIHgtbW9kZWw9InJ1bGUudGl0bGUiCiAgICAgICAgICAgIC8-CiAgICAgICAgICA8L2Rpdj4KICAgICAgICAgIDxkaXYgY2xhc3M9InJ1bGVfX2NlbGwgcnVsZV9fY2VsbC0tdHlwZSI-CiAgICAgICAgICAgIDxzZWxlY3QgeC1tb2RlbD0icnVsZS5jYXRlZ29yeSI-CiAgICAgICAgICAgICAgPG9wdGlvbiB2YWx1ZT0iYmxvY2siPkJsb2NrPC9vcHRpb24-CiAgICAgICAgICAgICAgPG9wdGlvbiB2YWx1ZT0iaW5saW5lIj5JbmxpbmU8L29wdGlvbj4KICAgICAgICAgICAgICA8b3B0aW9uIHZhbHVlPSJzZWxlY3RvciI-U2VsZWN0b3I8L29wdGlvbj4KICAgICAgICAgICAgPC9zZWxlY3Q-CiAgICAgICAgICA8L2Rpdj4KICAgICAgICAgIDxkaXYgY2xhc3M9InJ1bGVfX2NlbGwgcnVsZV9fY2VsbC0tc2VsZWN0b3IiPgogICAgICAgICAgICA8aW5wdXQKICAgICAgICAgICAgICB0eXBlPSJ0ZXh0IgogICAgICAgICAgICAgIHBsYWNlaG9sZGVyPSJlLmcuIGgxIgogICAgICAgICAgICAgIHgtbW9kZWw9InJ1bGUuc2VsZWN0b3IiCiAgICAgICAgICAgIC8-CiAgICAgICAgICA8L2Rpdj4KICAgICAgICAgIDxkaXYgY2xhc3M9InJ1bGVfX2NlbGwgcnVsZV9fY2VsbC0tY2xhc3NlcyI-CiAgICAgICAgICAgIDxpbnB1dAogICAgICAgICAgICAgIHR5cGU9InRleHQiCiAgICAgICAgICAgICAgcGxhY2Vob2xkZXI9ImUuZy4gaGVhZGluZy1wcmltYXJ5IgogICAgICAgICAgICAgIHgtbW9kZWw9InJ1bGUuY2xhc3NlcyIKICAgICAgICAgICAgLz4KICAgICAgICAgIDwvZGl2PgogICAgICAgICAgPGJ1dHRvbgogICAgICAgICAgICBjbGFzcz0icnVsZV9fZGVsZXRlIgogICAgICAgICAgICBAY2xpY2s9InJlbW92ZVJ1bGUoaW5kZXgpIgogICAgICAgICAgPgogICAgICAgICAgICBYCiAgICAgICAgICA8L2J1dHRvbj4KICAgICAgICAgIDxidXR0b24gY2xhc3M9InJ1bGVfX3NvcnQiPlNvcnQ8L2J1dHRvbj4KICAgICAgICA8L2Rpdj4KICAgICAgPC90ZW1wbGF0ZT4KICAgICAgPHRlbXBsYXRlIHgtaWY9InJ1bGUudHlwZSA9PT0gJ2dyb3VwJyI-CiAgICAgICAgPGRpdiBjbGFzcz0icnVsZSBydWxlLS1ncm91cCI-CiAgICAgICAgICA8ZGl2IGNsYXNzPSJydWxlX19jZWxsIHJ1bGVfX2NlbGwtLWdyb3VwLXRpdGxlIj4KICAgICAgICAgICAgPGlucHV0CiAgICAgICAgICAgICAgdHlwZT0idGV4dCIKICAgICAgICAgICAgICBwbGFjZWhvbGRlcj0iZS5nLiBIZWFkaW5ncyIKICAgICAgICAgICAgICB4LW1vZGVsPSJydWxlLnRpdGxlIgogICAgICAgICAgICAvPgogICAgICAgICAgPC9kaXY-CgogICAgICAgICAgPHRlbXBsYXRlCiAgICAgICAgICAgIHgtZm9yPSIoZ3JvdXBSdWxlLCBncm91cFJ1bGVJbmRleCkgaW4gcnVsZS5ncm91cFJ1bGVzIgogICAgICAgICAgICA6a2V5PSJncm91cFJ1bGVJbmRleCIKICAgICAgICAgID4KICAgICAgICAgICAgPGRpdiBjbGFzcz0icnVsZV9fY2VsbCBydWxlX19jZWxsLS1ydWxlcyI-CiAgICAgICAgICAgICAgPGRpdiBjbGFzcz0icnVsZSBydWxlLS1zaW5nbGUiPgogICAgICAgICAgICAgICAgPGRpdiBjbGFzcz0icnVsZV9fY2VsbCBydWxlX19jZWxsLS10aXRsZSI-CiAgICAgICAgICAgICAgICAgIDxpbnB1dAogICAgICAgICAgICAgICAgICAgIHR5cGU9InRleHQiCiAgICAgICAgICAgICAgICAgICAgcGxhY2Vob2xkZXI9InRpdGxlIgogICAgICAgICAgICAgICAgICAgIHgtbW9kZWw9Imdyb3VwUnVsZS50aXRsZSIKICAgICAgICAgICAgICAgICAgLz4KICAgICAgICAgICAgICAgIDwvZGl2PgogICAgICAgICAgICAgICAgPGRpdiBjbGFzcz0icnVsZV9fY2VsbCBydWxlX19jZWxsLS10eXBlIj4KICAgICAgICAgICAgICAgICAgPHNlbGVjdCB4LW1vZGVsPSJncm91cFJ1bGUuY2F0ZWdvcnkiPgogICAgICAgICAgICAgICAgICAgIDxvcHRpb24gdmFsdWU9ImJsb2NrIj5CbG9jazwvb3B0aW9uPgogICAgICAgICAgICAgICAgICAgIDxvcHRpb24gdmFsdWU9ImlubGluZSI-SW5saW5lPC9vcHRpb24-CiAgICAgICAgICAgICAgICAgICAgPG9wdGlvbiB2YWx1ZT0ic2VsZWN0b3IiPgogICAgICAgICAgICAgICAgICAgICAgU2VsZWN0b3IKICAgICAgICAgICAgICAgICAgICA8L29wdGlvbj4KICAgICAgICAgICAgICAgICAgPC9zZWxlY3Q-CiAgICAgICAgICAgICAgICA8L2Rpdj4KICAgICAgICAgICAgICAgIDxkaXYKICAgICAgICAgICAgICAgICAgY2xhc3M9InJ1bGVfX2NlbGwgcnVsZV9fY2VsbC0tc2VsZWN0b3IiCiAgICAgICAgICAgICAgICA-CiAgICAgICAgICAgICAgICAgIDxpbnB1dAogICAgICAgICAgICAgICAgICAgIHR5cGU9InRleHQiCiAgICAgICAgICAgICAgICAgICAgcGxhY2Vob2xkZXI9InNlbGVjdG9yIgogICAgICAgICAgICAgICAgICAgIHgtbW9kZWw9Imdyb3VwUnVsZS5zZWxlY3RvciIKICAgICAgICAgICAgICAgICAgLz4KICAgICAgICAgICAgICAgIDwvZGl2PgogICAgICAgICAgICAgICAgPGRpdiBjbGFzcz0icnVsZV9fY2VsbCBydWxlX19jZWxsLS1jbGFzc2VzIj4KICAgICAgICAgICAgICAgICAgPGlucHV0CiAgICAgICAgICAgICAgICAgICAgdHlwZT0idGV4dCIKICAgICAgICAgICAgICAgICAgICBwbGFjZWhvbGRlcj0iY2xhc3NlcyIKICAgICAgICAgICAgICAgICAgICB4LW1vZGVsPSJncm91cFJ1bGUuY2xhc3NlcyIKICAgICAgICAgICAgICAgICAgLz4KICAgICAgICAgICAgICAgIDwvZGl2PgogICAgICAgICAgICAgICAgPGJ1dHRvbgogICAgICAgICAgICAgICAgICBjbGFzcz0icnVsZV9fZGVsZXRlIgogICAgICAgICAgICAgICAgICBAY2xpY2s9InJlbW92ZUdyb3VwUnVsZShpbmRleCwgZ3JvdXBSdWxlSW5kZXgpIgogICAgICAgICAgICAgICAgPgogICAgICAgICAgICAgICAgICBYCiAgICAgICAgICAgICAgICA8L2J1dHRvbj4KICAgICAgICAgICAgICA8L2Rpdj4KICAgICAgICAgICAgPC9kaXY-CiAgICAgICAgICA8L3RlbXBsYXRlPgogICAgICAgICAgPGRpdiBjbGFzcz0iYWRkLXJ1bGVzIj4KICAgICAgICAgICAgPGJ1dHRvbgogICAgICAgICAgICAgIGNsYXNzPSJhZGQtcnVsZXNfX2J0biBidG4ganMtYWRkLXN0eWxlLXRvLWdyb3VwIgogICAgICAgICAgICAgIEBjbGljaz0iYWRkUnVsZVRvR3JvdXAoaW5kZXgpIgogICAgICAgICAgICA-CiAgICAgICAgICAgICAgQWRkIFN0eWxlIFRvIEdyb3VwCiAgICAgICAgICAgIDwvYnV0dG9uPgogICAgICAgICAgPC9kaXY-CgogICAgICAgICAgPGJ1dHRvbgogICAgICAgICAgICBjbGFzcz0icnVsZV9fZGVsZXRlIgogICAgICAgICAgICBAY2xpY2s9InJlbW92ZVJ1bGUoaW5kZXgpIgogICAgICAgICAgPgogICAgICAgICAgICBYCiAgICAgICAgICA8L2J1dHRvbj4KICAgICAgICAgIDxidXR0b24gY2xhc3M9InJ1bGVfX3NvcnQiPlNvcnQ8L2J1dHRvbj4KICAgICAgICA8L2Rpdj4KICAgICAgPC90ZW1wbGF0ZT4KICAgIDwvZGl2PgogIDwvdGVtcGxhdGU-CgogIDxkaXYgY2xhc3M9ImFkZC1ydWxlcyI-CiAgICA8YnV0dG9uCiAgICAgIEBjbGljaz0iYWRkUnVsZSgnc2luZ2xlJykiCiAgICAgIGNsYXNzPSJhZGQtcnVsZXNfX2J0biBidG4iCiAgICA-CiAgICAgIEFkZCBTaW5nbGUgU3R5bGUKICAgIDwvYnV0dG9uPgogICAgPGJ1dHRvbgogICAgICBAY2xpY2s9ImFkZFJ1bGUoJ2dyb3VwJykiCiAgICAgIGNsYXNzPSJhZGQtcnVsZXNfX2J0biBidG4iCiAgICA-CiAgICAgIEFkZCBTdHlsZSBHcm91cAogICAgPC9idXR0b24-CiAgPC9kaXY-CgogIDxkaXYgY2xhc3M9Im91dHB1dCI-CiAgICA8cD5PdXRwdXQ6PC9wPgogICAgPHRleHRhcmVhCiAgICAgIHJlYWRvbmx5CiAgICAgIHgtdGV4dD0ib3V0cHV0IgogICAgICBjbGFzcz0ib3V0cHV0LXVpX190ZXh0YXJlYSBqcy1vdXRwdXQiCiAgICA-PC90ZXh0YXJlYT4KICA8L2Rpdj4KPC9kaXY-Cg&script=aW1wb3J0IFNvcnRhYmxlLCB7IFNvcnRhYmxlRXZlbnQgfSBmcm9tICdzb3J0YWJsZWpzJzsKCmludGVyZmFjZSBCYXNlUnVsZSB7CiAgdGl0bGU6IHN0cmluZzsKICBjYXRlZ29yeTogJ2Jsb2NrJyB8ICdpbmxpbmUnIHwgJ3NlbGVjdG9yJzsKICBzZWxlY3Rvcjogc3RyaW5nOwogIGNsYXNzZXM6IHN0cmluZzsKfQoKaW50ZXJmYWNlIFNpbmdsZVJ1bGUgZXh0ZW5kcyBCYXNlUnVsZSB7CiAgdHlwZTogJ3NpbmdsZSc7Cn0KCmludGVyZmFjZSBHcm91cFJ1bGUgZXh0ZW5kcyBCYXNlUnVsZSB7fQoKaW50ZXJmYWNlIEdyb3VwIGV4dGVuZHMgQmFzZVJ1bGUgewogIHR5cGU6ICdncm91cCc7CiAgZ3JvdXBSdWxlczogR3JvdXBSdWxlW107Cn0KCnR5cGUgUnVsZSA9IFNpbmdsZVJ1bGUgfCBHcm91cDsKCmxldCBpZCA9IDE7CgpBbHBpbmUuZGF0YSgKICAncnVsZUNvbmZpZ3VyYXRvcicsCiAgZnVuY3Rpb24gcnVsZUNvbmZpZ3VyYXRvcigpIHsKICAgIHJldHVybiB7CiAgICAgIHJ1bGVzOiBbXSBhcyBSdWxlW10sCiAgICAgIGdldCBvdXRwdXQoKSB7CiAgICAgICAgcmV0dXJuIEpTT04uc3RyaW5naWZ5KHRoaXMucnVsZXMsIG51bGwsIDIpOwogICAgICB9LAoKICAgICAgaW5pdCgpIHsKICAgICAgICB0aGlzLmluaXRTb3J0YWJsZSgpOwogICAgICB9LAoKICAgICAgaW5pdFNvcnRhYmxlKCkgewogICAgICAgIGNvbnN0IHJ1bGVzQ29udGFpbmVyID0gZG9jdW1lbnQucXVlcnlTZWxlY3RvcigKICAgICAgICAgICcuY29uZmlndXJhdG9yJywKICAgICAgICApIGFzIEhUTUxFbGVtZW50OwogICAgICAgIFNvcnRhYmxlLmNyZWF0ZShydWxlc0NvbnRhaW5lciwgewogICAgICAgICAgaGFuZGxlOiAnLnJ1bGVfX3NvcnQnLAogICAgICAgICAgb25FbmQ6IChldmVudDogU29ydGFibGVFdmVudCkgPT4gewogICAgICAgICAgICBjb25zb2xlLmxvZyhldmVudCk7CiAgICAgICAgICAgIGNvbnNvbGUubG9nKHRoaXMub3V0cHV0KTsKICAgICAgICAgICAgY29uc3QgbW92ZWRSdWxlID0gdGhpcy5ydWxlcy5zcGxpY2UoCiAgICAgICAgICAgICAgZXZlbnQub2xkSW5kZXghLAogICAgICAgICAgICAgIDEsCiAgICAgICAgICAgIClbMF07CiAgICAgICAgICAgIHRoaXMucnVsZXMuc3BsaWNlKAogICAgICAgICAgICAgIGV2ZW50Lm5ld0luZGV4ISwKICAgICAgICAgICAgICAwLAogICAgICAgICAgICAgIG1vdmVkUnVsZSwKICAgICAgICAgICAgKTsKICAgICAgICAgICAgY29uc29sZS5sb2codGhpcy5vdXRwdXQpOwogICAgICAgICAgfSwKICAgICAgICB9KTsKICAgICAgfSwKCiAgICAgIGFkZFJ1bGUodHlwZTogJ3NpbmdsZScgfCAnZ3JvdXAnKSB7CiAgICAgICAgaWYgKHR5cGUgPT09ICdzaW5nbGUnKSB7CiAgICAgICAgICB0aGlzLnJ1bGVzLnB1c2goewogICAgICAgICAgICBpZDogaWQrKywKICAgICAgICAgICAgdHlwZTogJ3NpbmdsZScsCiAgICAgICAgICAgIHRpdGxlOiAnJywKICAgICAgICAgICAgY2F0ZWdvcnk6ICdibG9jaycsCiAgICAgICAgICAgIHNlbGVjdG9yOiAnJywKICAgICAgICAgICAgY2xhc3NlczogJycsCiAgICAgICAgICB9IGFzIFNpbmdsZVJ1bGUpOwogICAgICAgIH0gZWxzZSB7CiAgICAgICAgICB0aGlzLnJ1bGVzLnB1c2goewogICAgICAgICAgICBpZDogaWQrKywKICAgICAgICAgICAgdHlwZTogJ2dyb3VwJywKICAgICAgICAgICAgdGl0bGU6ICcnLAogICAgICAgICAgICBjYXRlZ29yeTogJ2Jsb2NrJywKICAgICAgICAgICAgc2VsZWN0b3I6ICcnLAogICAgICAgICAgICBjbGFzc2VzOiAnJywKICAgICAgICAgICAgZ3JvdXBSdWxlczogWwogICAgICAgICAgICAgIHsKICAgICAgICAgICAgICAgIGlkOiBpZCsrLAogICAgICAgICAgICAgICAgdGl0bGU6ICcnLAogICAgICAgICAgICAgICAgY2F0ZWdvcnk6ICdibG9jaycsCiAgICAgICAgICAgICAgICBzZWxlY3RvcjogJycsCiAgICAgICAgICAgICAgICBjbGFzc2VzOiAnJywKICAgICAgICAgICAgICB9LAogICAgICAgICAgICBdLAogICAgICAgICAgfSBhcyBHcm91cCk7CiAgICAgICAgfQogICAgICB9LAoKICAgICAgYWRkUnVsZVRvR3JvdXAocnVsZUluZGV4OiBudW1iZXIpIHsKICAgICAgICBjb25zdCBydWxlID0gdGhpcy5ydWxlc1tydWxlSW5kZXhdOwogICAgICAgIGlmIChydWxlLnR5cGUgPT09ICdncm91cCcpIHsKICAgICAgICAgIHJ1bGUuZ3JvdXBSdWxlcy5wdXNoKHsKICAgICAgICAgICAgdGl0bGU6ICcnLAogICAgICAgICAgICBjYXRlZ29yeTogJ2Jsb2NrJywKICAgICAgICAgICAgc2VsZWN0b3I6ICcnLAogICAgICAgICAgICBjbGFzc2VzOiAnJywKICAgICAgICAgIH0pOwogICAgICAgIH0KICAgICAgfSwKCiAgICAgIHJlbW92ZVJ1bGUoaW5kZXg6IG51bWJlcikgewogICAgICAgIHRoaXMucnVsZXMuc3BsaWNlKGluZGV4LCAxKTsKICAgICAgfSwKCiAgICAgIHJlbW92ZUdyb3VwUnVsZSgKICAgICAgICBydWxlSW5kZXg6IG51bWJlciwKICAgICAgICBncm91cFJ1bGVJbmRleDogbnVtYmVyLAogICAgICApIHsKICAgICAgICBjb25zdCBydWxlID0gdGhpcy5ydWxlc1tydWxlSW5kZXhdOwogICAgICAgIGlmIChydWxlLnR5cGUgPT09ICdncm91cCcpIHsKICAgICAgICAgIHJ1bGUuZ3JvdXBSdWxlcy5zcGxpY2UoZ3JvdXBSdWxlSW5kZXgsIDEpOwogICAgICAgIH0KICAgICAgfSwKICAgIH07CiAgfSwKKTsK)

So the changes were basically

1. add unique ids to rules
2. use a reactive getter for the output and stop trying to manually update it

**Reply by danbramall** [reply]
Thanks! I did try the core Sort plugin but came across a bug with the way the sort handles work and also couldn't figure out how to update the rules json object. It's my first Alpine project so a bit of a learning exercise. Thanks so much for your pointers :)

**Reply by danbramall** [reply]
It seems to be that this doesn't work on initial load but if I change the typescript file (e.g. comment out one of the console.logs) it rebuilds and then does work. Having pulled this code down to my local repo I can only get it in the initial state, i.e. the json output updates correctly but the UI immediately reverts after sorting.

**Reply by danbramall** [reply]
I'm currently going down this wormhole https://github.com/alpinejs/alpine/discussions/1635. It seems to work using `Alpine.Raw` and updating `_x_prevKeys` in the sort method, but feels hacky.

**Reply by ekwoka** [reply]
hmm yeah, there are some issues.

btw, you don't need to do `x-init="init"` `init()` is automatically called when the component is made. in your tests, that may also cause some issues.
